src/playground-project.ts

```diff
@@ -223,79 +263,48 @@ export class PlaygroundProject extends LitElement {
     this._compileProject();
   }
 
-  private async _startWorkers() {
-    await Promise.all([
-      this._startTypeScriptWorker(),
-      this._installServiceWorker(),
-    ]);
+  firstUpdated() {
+    const worker = new Worker(typescriptWorkerScriptUrl);
+    this._typescriptWorkerAPI = wrap<TypeScriptWorkerAPI>(worker);
   }
 
-  private async _startTypeScriptWorker() {
-    if (this._typescriptWorkerAPI === undefined) {
-      const worker = new Worker(typescriptWorkerScriptUrl);
-      this._typescriptWorkerAPI = wrap<TypeScriptWorkerAPI>(worker);
-    } else {
-      console.debug('typescript-worker already started');
-    }
-  }
-
-  private async _installServiceWorker() {
-    if (!('serviceWorker' in navigator)) {
-      // TODO: show this in the UI
-      console.warn('ServiceWorker support required for <playground-project>');
+  private _onServiceWorkerProxyIframeLoad() {
+    const window = this._iframe.contentWindow;
+    if (!window) {
+      console.error('IFrame did not have window');
       return;
     }
-
-    const registration = await navigator.serviceWorker.register(
-      serviceWorkerScriptUrl.href,
-      {scope: this._scopeUrl}
+    // This channel is persistent, and is only used to receieve new service
+    // worker channel ports from the proxy iframe. Note we can get new service
+    // worker ports at any time from the proxy, when the service worker updates.
+    const {port1, port2} = new MessageChannel();
+    port1.addEventListener('message', (event: MessageEvent<MessagePort>) =>
+      this._onNewServiceWorkerPort(event.data)
     );
-
-    registration.addEventListener('updatefound', () => {
-      // We can get a new service worker at any time, so we need to listen for
-      // updates and connect to new workers on demand.
-      const newWorker = registration.installing;
-      if (newWorker) {
-        this._connectServiceWorker(newWorker);
-      }
-    });
-
-    if (registration.active) {
-      this._connectServiceWorker(registration.active);
-    } else {
-      console.warn('unhandled service worker registration state', registration);
-    }
+    port1.start();
+    const initMessage: ProxyInitMessage = {
+      port: port2,
+      url: 'playground-service-worker.js',
+      scope: this.sandboxScope,
+    };
+    window.postMessage(initMessage, '*', [initMessage.port]);
```

Comment:
We can do better than `'*'` now, right? Does it matter?

Reply:
We could constrain it to the configured origin, but that would add the limitation that the URL may not redirect to another origin. I'm not sure what security issue this would protect against, and could be an inconvenience. It seems like if an attacker was able to redirect that URL somewhere else, then all bets are off anyway?